extensions/positron-r/src/provider.ts

```diff
@@ -25,123 +25,101 @@ export const R_DOCUMENT_SELECTORS = [
+	const systemHqBinaries = discoverHQBinaries(rHeadquarters());
+	updateBinaries(systemHqBinaries);
+
+	const registryBinaries = await discoverRegistryBinaries();
+	updateBinaries(registryBinaries);
 
-	// other conventional places we might find an R binary (or a symlink to one)
-	const possibleBinaries = [
+	const moreBinaries = discoverAdHocBinaries([
 		'/usr/bin/R',
 		'/usr/local/bin/R',
 		'/opt/local/bin/R',
 		'/opt/homebrew/bin/R'
-	];
-	const moreBinaries = possibleBinaries
-		.filter(b => fs.existsSync(b))
-		.map(b => fs.realpathSync(b));
-	for (const b of moreBinaries) {
-		if (!binaries.has(b)) {
-			binaries.set(b, BinarySource.adHoc);
-		}
-	}
-
-	// same as above but user-specified, ad hoc binaries
-	const userPossibleBinaries = userRBinaries();
-	const userMoreBinaries = userPossibleBinaries
-		.filter(b => fs.existsSync(b))
-		.map(b => fs.realpathSync(b));
-	for (const b of userMoreBinaries) {
-		if (!binaries.has(b)) {
-			binaries.set(b, BinarySource.user);
-		}
-	}
+	]);
+	updateBinaries(moreBinaries);
 
-	const registryBinaries = await discoverRegistryBinaries();
-	for (const b of registryBinaries) {
-		if (!binaries.has(b)) {
-			binaries.set(b, BinarySource.registry);
-		}
-	}
+	// Optional, user-specified root directories or binaries
+	const userHqBinaries = discoverHQBinaries(userRHeadquarters());
+	updateBinaries(userHqBinaries);
 
-	const pathBinary = await findRBinaryFromPATH();
-	if (pathBinary && !binaries.has(pathBinary)) {
-		binaries.set(pathBinary, BinarySource.PATH);
-	}
+	const userMoreBinaries = discoverAdHocBinaries(userRBinaries());
+	updateBinaries(userMoreBinaries);
 
-	// make sure we include the "current" version of R, for some definition of "current"
-	// we've probably already discovered it, but we still want to single it out, so that we mark
-	// that particular R installation as the current one
-	const curBin = await findCurrentRBinary();
-	if (curBin) {
-		rInstallations.push(new RInstallation(curBin, true));
-		binaries.delete(curBin);
+	// (Try to) promote each RBinary to a proper RInstallation
+	let rInstallations: Array<RInstallation> = [];
+	if (currentBinaries.length > 0) {
+		const currentBinary = currentBinaries[0];
+		rInstallations.push(new RInstallation(currentBinary.path, true, currentBinary.reasons));
+		binaries.delete(currentBinary.path);
 	}
 
-	binaries.forEach((source, bin) => {
-		rInstallations.push(new RInstallation(bin));
+	binaries.forEach(rbin => {
+		rInstallations.push(new RInstallation(rbin.path, false, rbin.reasons));
 	});
 
-	// TODO: possible location to tell the user why certain R installations are being omitted from
-	// the interpreter drop-down and, in some cases, offer to help fix the situation:
-	// * version < minimum R version supported by positron-r
-	// * (macOS only) version is not orthogonal and is not the current version of R
-	// * invalid R installation
+	const rejectedRInstallations: Array<RInstallation> = [];
```

Comment:
Starting to keep the rejected R installations around, at least for a little while. This lays the groundwork for possibly incorporating this info into a section of a picker (see #5223) that could be hidden/unhidden.